nlp/2021_11_25_gpt3mix/nlp_gpt3mix.ipynb

```diff
@@ -0,0 +1,1966 @@
+        "\n",
+        "We will generate 10, 50, 100 and 200 synthetic samples using GPT-3 to investigate the effectiveness of text augmentation."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "BAzLn9t83YCS"
+      },
+      "source": [
+        "# define the number of synthetic samples to generate\n",
+        "n = 10\n",
+        "new_texts = []\n",
+        "new_labels = []\n",
+        "api_key =  # insert your api key for GPT-3\n",
+        "headers = {'Authorization' : 'Bearer ' + api_key ,\n",
+        "              'Content-type':'application/json', \n",
+        "              'Accept':'application/json'}\n",
+        "\n",
+        "iter = 0\n",
+        "while iter < n:\n",
+        "  # select two random samples from training set\n",
+        "  text1, label1, text2, label2 = get_random_samples()\n",
+        "  # create the prompt\n",
+        "  prompt = get_prompt(text1, label1, text2, label2)\n",
+        "  # send a post request to gpt-3 using the prompt\n",
+        "  response = requests.post('https://api.openai.com/v1/engines/davinci/completions', \n",
+        "                           headers=headers,\n",
+        "                           data = json.dumps({\"prompt\": prompt, \n",
+        "                                              \"max_tokens\": 30,\n",
+        "                                              \"temperature\": 0.9,\n",
+        "                                              \"top_p\": 0.95}))\n",
+        "\n",
+        "  # get response and extract the generated text and label\n",
+        "  # the generated output will be in the form \"<text> (Sentiment: <label>)\"\n",
+        "  data = response.json()['choices'][0]['text'].split('\\n')[0].split('(Sentiment:')\n",
+        "\n",
+        "  if len(data) < 2:\n",
+        "    # the format of the response is invalid\n",
+        "    continue\n",
+        "\n",
+        "  text = data[0]\n",
+        "  label = data[1].split(')')[0].strip()\n",
+        "\n",
+        "  if label not in ['joy', 'anger', 'surprise']:\n",
+        "    # the format of the response is invalid\n",
+        "    continue\n",
+        "\n",
+        "  new_texts.append(text)\n",
+        "  new_labels.append(label2idx[label])\n",
+        "  iter += 1\n",
+        "\n",
+        "# define the synthetic dataset and save it to disk so as to prevent sending \n",
+        "# many api requests\n",
+        "synthetic_ds = Dataset.from_dict({'text': new_texts, 'label': new_labels})\n",
+        "synthetic_ds.save_to_disk('./drive/MyDrive/text_augmentation/gpt-3/10')"
+      ],
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "rPlF7dXKor6v"
+      },
+      "source": [
```

Comment:
I would re-work this bit and host these files in the Github folder, and this cell fetches them through a `git clone` action. Just because not everybody will have access to GPT3, but they might just wanna run the notebook either way. Like we did in this tip: https://github.com/ml6team/quick-tips/tree/main/nlp/2021_06_29_long_range_transformers